### ISSUE TITLE
Android App automatically logs out after few hours of inactivity

### ISSUE THREAD

**Issue by johnkj** [owncloud/android#1964]
### Actual behaviour
The latest version of android app 2.3.0 with little customization installs as usual but:
1-when connected to an account on owncloud 10 with php 7/7.1/5.6 automatically logs out after few hours of inactivity on the app and after that even if you give the correct password it will not login unless you uninstall the app and reinstall it.
2-When connected to an account on owncloud 9.1.0 with php 5.6/5.5 it does not log out. 
This occurs when the app is installed from playstore only, this does not happen when installed directly.

### Expected behaviour
Once logged into the app, it should remain logged in.
 
### Steps to reproduce
1. Build owncloud android 2.3.0 with basic changes in logo and setup.xml
2. upload to playstore and install on android phone from playstore 
3. Add owncloud account which is running on OC 10 and another account running on OC 9.1.0
4- It will login successfully, and after few hours about 3-4 hours of inactivity.
5- Open the owncloud android app, it will logout and redirect to login screen with server url and username prefilled asking you to enter the password
6- Even if you enter the correct password it will not accept.
7-But adding a new account with different username will work.

Can this problem be reproduced with the official owncloud server?
(url: https://demo.owncloud.org, user: test, password: test)

### Environment data
Android version: 6.0.1, 5.1

Device model: 

Stock or customized system:

ownCloud app version: 2.3.0

ownCloud server version: 10.0.0.12

#### Web server error log
```
Insert your webserver log here
42.109.141.168 - testyg [12/May/2017:10:30:00 +0000] "GET /ocs/v1.php/cloud/capabilities?format=json HTTP/1.1" 200 4662 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - testyg [12/May/2017:10:30:00 +0000] "GET /ocs/v1.php/cloud/user?format=json HTTP/1.1" 200 864 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - testyg [12/May/2017:10:30:01 +0000] "GET /index.php/avatar/testyg/256 HTTP/1.1" 200 784 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - testyg [12/May/2017:10:30:01 +0000] "PROPFIND /remote.php/webdav/ HTTP/1.1" 401 811 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - - [12/May/2017:10:30:01 +0000] "GET /status.php HTTP/1.1" 200 1183 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - - [12/May/2017:10:30:01 +0000] "PROPFIND /remote.php/webdav/ HTTP/1.1" 401 1315 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - - [12/May/2017:10:30:02 +0000] "GET /status.php HTTP/1.1" 200 1183 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"
42.109.141.168 - - [12/May/2017:10:30:02 +0000] "PROPFIND /remote.php/webdav/ HTTP/1.1" 401 1309 "-" "Mozilla/5.0 (Android) owncloud-android/2.3.0"

#### ownCloud log (data/owncloud.log)
``
Insert your ownCloud log here
{"reqId":"XNTwmVMw6Qnm8cHhHwfj","level":0,"time":"2017-05-12T16:00:01+05:30","remoteAddr":"42.109.141.168","user":"testyg","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 CSRF check not passed.\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/html\\\/backup\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Auth.php(146): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Auth->auth(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(202): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Auth->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(150): Sabre\\\\DAV\\\\Auth\\\\Plugin->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#6 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/var\\\/www\\\/html\\\/backup\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/var\\\/www\\\/html\\\/backup\\\/remote.php(165): require_once('\\\/var\\\/www\\\/html\\\/b...')\\n#9 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/backup\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Auth.php\",\"Line\":211,\"User\":\"testyg\"}"}
{"reqId":"p1Ry9EjrauB8MP8rPi3t","level":0,"time":"2017-05-12T16:00:01+05:30","remoteAddr":"42.109.141.168","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/html\\\/backup\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/var\\\/www\\\/html\\\/backup\\\/remote.php(165): require_once('\\\/var\\\/www\\\/html\\\/b...')\\n#6 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php\",\"Line\":168,\"User\":false}"}
{"reqId":"NZ8NSNFjMoZOo8iWUd51","level":0,"time":"2017-05-12T16:00:02+05:30","remoteAddr":"42.109.141.168","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/var\\\/www\\\/html\\\/backup\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/var\\\/www\\\/html\\\/backup\\\/remote.php(165): require_once('\\\/var\\\/www\\\/html\\\/b...')\\n#6 {main}\",\"File\":\"\\\/var\\\/www\\\/html\\\/backup\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php\",\"Line\":168,\"User\":false}"}

**Comment by davigonz** [comment]
Hi @johnkj , thank you for reporting this problem. Could you please provide us more information about your server configuration, PHP version for instance, just to discard that the wrong behavior is not related with the server and is an app issue?

**Comment by DaanTry** [comment]
I have exact the same issue.

Android version: 7.0
Device model: SM-G935F
Stock or customized system: Stock
ownCloud app version: 2.5.0
ownCloud server version: 10.0.4 (on PHP 7.0.16)

Only removing and reinstalling App from Appstore resolves the issue.

**Comment by jensd0e** [comment]
I guess, this isn't a new issue. Some people (including me) reported this before.

**Comment by jesmrec** [comment]
we have the following opened with the same issue:

https://github.com/owncloud/android/issues/1964

We can keep on discussing there. If you think is a different issue, please reopen

**Comment by davigonz** [comment]
@jensd0e , @jesmrec are you sure this is the same issue that #2030? 

@johnkj Are you using LDAP? Are you using spaces in your username? Do you notice this behaviour described by @jensd0e in the other issue?

> Username is changed from 'Firstname Lastname' to 'Firstname_Lastname' and the username field can't be modified, so the username can't be changed and the user can't authenticate against LDAP server.

**Comment by jesmrec** [comment]
that issue refers directly to ldap or blanks but i mean that the problem is not in this set-up. Bug appears regardless LDAP is enabled or not

**Comment by jensd0e** [comment]
@davigonz I'm not that sure anymore. There where two problems. Username normalization and being redirected to login screen. If username normalization would cause my problems, I would see logon errors on my windows server. But for me, logon is ok and only after login I'm redirected to login screen. In #2030 I started with problem No. 1 (username normalization) and continued with problem No. 2 (being redirected). That was a bad idea, I guess.

**Comment by jesmrec** [comment]
ok, we reopen and keep the discussion alive

**Comment by DaanTry** [comment]
In my case no spaces in de username and no LDAP used for authentication. Reproducability is 100%. A few days after installing the issue appears and can only be resolved by uninstalling/reinstalling the app.

**Comment by davigonz** [comment]
Hi @DaanTry , thank you for your feedback, can you create a test account and send us the credentials to [apps@owncloud.com](apps@owncloud.com) so that we can try to reproduce it? Thanks

**Comment by jesmrec** [comment]
@DaanTry if you send the credentials, please include a link to this issue within the email.

**Comment by jesmrec** [comment]
@DaanTry credentials received, thanks a lot!!!!

**Comment by gfredericks** [comment]
This has been happening to me for a week or two now. It also seems to always take exactly two logins to get it working, for some reason.

**Comment by davigonz** [comment]
Hi @gfredericks , what's your ownCloud app version and which device are you using? Thanks

**Comment by gfredericks** [comment]
android, 2.5.0

**Comment by davigonz** [comment]
> Add owncloud account which is running on OC 10 and another account running on OC 9.1.0
4- It will login successfully, and after few hours about 3-4 hours of inactivity.

@johnkj which of these accounts is redirected to login view? Can you try out if this happens with both accounts?

**Comment by davigonz** [comment]
Another question @johnkj , are you using http or https?

**Comment by davigonz** [comment]
Hi @johnkj, I've managed to reproduce your problem and seems to be related with your server version (9.1.0) because of a bug in ownCloud server (https://github.com/owncloud/core/issues/30196)

I suggest you to update your server to the stable version of 9.1, which is in this moment 9.1.7 and try to have 9.1.7 account along with 10 and see what happens.

@DaanTry Your case is different, you are using 10.0.4 server version and we are not able to reproduce your problem, even using your credentials. Can you please download our [beta](https://play.google.com/store/apps/details?id=com.owncloud.android.beta&hl=es), check if the error persists and send us some logs (Side menu-> Settings -> Logs) to [apps@owncloud.com](apps@owncloud.com) when the redirection appears.

**Comment by PVince81** [comment]
@davigonz my wife just go this CSRF thing with 10.0.5 final which I had on my server for quick testing before announcing.

She has 2.5.1-beta.

Since I upgraded from 10.0.5 RC3 to 10.0.5 final today the problem appeared, not sure why. There is no code difference between these versions and the Android app worked fine before that. Maybe her session got killed by the update somehow ?

Now the app is repeatedly asking for her password and when she enters it, the following logs are produced:

This is the log:
```
{"reqId":"WmjpFX8AAAIAAEbEzVkAAAAD","level":0,"time":"2018-01-24T20:14:13+00:00","remoteAddr":"77.178.180.25","user":"tingting","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 CSRF check not passed.\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Auth.php(150): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Auth->auth(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(199): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Auth->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php(150): Sabre\\\\DAV\\\\Auth\\\\Plugin->check(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(165): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Auth.php\",\"Line\":215,\"User\":\"tingting\"}"}
{"reqId":"WmjpFX8AAAIAAEbEzVsAAAAD","level":0,"time":"2018-01-24T20:14:14+00:00","remoteAddr":"77.178.180.25","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(165): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#6 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php\",\"Line\":168,\"User\":false}"}
{"reqId":"WmjpFn8AAAIAAEbEzV0AAAAD","level":0,"time":"2018-01-24T20:14:14+00:00","remoteAddr":"77.178.180.25","user":"--","app":"webdav","method":"PROPFIND","url":"\/remote.php\/webdav\/","message":"Exception: {\"Message\":\"HTTP\\\/1.1 401 No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\NotAuthenticated\",\"Code\":0,\"Trace\":\"#0 [internal function]: Sabre\\\\DAV\\\\Auth\\\\Plugin->beforeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(466): Sabre\\\\Event\\\\EventEmitter->emit('beforeMethod', Array)\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(63): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(165): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#6 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Auth\\\/Plugin.php\",\"Line\":168,\"User\":false}"}
```

Need to investigate also server side.

**Comment by PVince81** [comment]
This CSRF thing doesn't even make sense for the mobile app ?! Or is it using some kind of web view ?

**Comment by DaanTry** [comment]
I have temporary disabled the csrf check in config.php. That solves the issue with the App but it is not preferred of course.
-------- Original message --------From: Vincent Petry <notifications@github.com> Date: 1/24/18  21:21  (GMT+01:00) To: owncloud/android <android@noreply.github.com> Cc: DaanTry <Daan.Try@GMail.com>, Mention <mention@noreply.github.com> Subject: Re: [owncloud/android] Android App automatically logs out after few hours of inactivity (#1964) 
This CSRF thing doesn't even make sense for the mobile app ?! Or is it using some kind of web view ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/owncloud/android","title":"owncloud/android","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/owncloud/android"}},"updates":{"snippets":[{"icon":"PERSON","message":"@PVince81 in #1964: This CSRF thing doesn't even make sense for the mobile app ?! Or is it using some kind of web view ?"}],"action":{"name":"View Issue","url":"https://github.com/owncloud/android/issues/1964#issuecomment-360261101"}}}

**Comment by PVince81** [comment]
I tried deleting all PHP session files but somehow it didn't kick out any people (likely another bug...)

I managed to restore my wife's account by manually deleting the matching entry in `oc_authtoken` table where the connection name is "Mozilla/5.0 (Android) ownCloudBeta-android/2.5.1-beta.1" and type 1. Now she can login again.

This also mean we'll likely not be able to reproduce the problem again on our end for debugging... I hope to get more light on how authentication is done from @davigonz and where that CSRF error might come from.

Since people had this in past versions of OC this is hopefully unrelated to any changes in 10.0.5.

**Comment by PVince81** [comment]
Additional info: my wife original used an app password. Then after today's server upgrade she lost the connection. Then she tried using her regular password. Not sure if that matters in this scenario, just pointing it out in case it reveals more clues.
This resulted in two separate entries in the oc_authtoken table. I deleted both to restore account access.

If the Android app is using a web view and/or is storing the CSRF value, it should delete it if it receives a 401 error and use the new one from the login page.

**Comment by SamuAlfageme** [comment]
@PVince81 do you happen to know if the app password was created at some point prior to the server being 10.0.4? there are some issues with them described in https://github.com/owncloud/core/issues/30157

**Comment by PVince81** [comment]
@SamuAlfageme I think I had 10.0.4 on my server when she created the app password last December.

Hmm, so that core bug tells that upgrading from OC < 10.0.4 to 10.0.4 or later would likely invalidate passwords ? Maybe that's the same bug. I'll see if I can reproduce this locally.

However it seems we now have two bugs:

1) the app password invalidation, which should not happen: https://github.com/owncloud/core/issues/30157

and

2) if app password was invalidated legitimately, the Android app should not get stuck with a CSRF error

Can you tell us more about how Android authenticates and CSRF ? It sounds like it's using the PHP login page in the background which is rather hacky. (a hack that might be legit if other ways like Webdav did not work in the past for some reason)

**Comment by davigonz** [comment]
> This CSRF thing doesn't even make sense for the mobile app ?! Or is it using some kind of web view ?

@PVince81 Thank you so much for your feeback.

The Android app is using a web view with SAML or OAuth authentication, is your wife using these authentication methods or just basic?

These are the steps the Android app follows when logging in: 

1. GET status.php to retrieve the server information.
2. PROPFIND to /remote.php/webdav to know which authentication methods are supported by the server.
      - 401: Basic => Android app shows the credentials fields (username & password) so that user 
                                can fill them.
                 OAuth => Open webview which will continue with the login process
      - 302: SAML => Open webview which will continue with the login process
3. In the case of basic, more authenticated requests to retrieve user data, capabilities, folders and so on.

If you need some mitmproxy logs, just ask me for them.

**Comment by davigonz** [comment]
> Is the Android client sending the CSRF request header with basic auth ? I don't see why it would we needed for basic auth.

I do not see CSRF header in our requests, at least in the login ones I described above. I will send you the logs of the login flow with the headers and content for the requests/responses, maybe you notice something weird there.

**Comment by PVince81** [comment]
Let's look at the server code: https://github.com/owncloud/core/blob/v10.0.4/apps/dav/lib/Connector/Sabre/Auth.php#L166

These are the conditions that would trigger a CSRF check for any DAV call:
- method is not GET
- user agent is neither desktop, android or ios
- user is logged in
- method is POST
- user is not "DAV authenticated" (not clear what this means, possibly that the user got authenticated over Webdav, not web UI)

Now I wonder whether we really need a CSRF check for PROPFIND... Probably not. (to be discussed separately)

The core problem is likely due to some of these conditions not met in that non-reproducible scenario.

**Comment by PVince81** [comment]
ohhh... this "DAV authenticated" condition is based on a session variable.

It is possible that the session was in a weird state: the user was still logged in (because if `isLoggedIn()` is false, the CSRF check is cancelled) but not DAV authenticated, maybe the session value is gone ?

Maybe something happened to the session and/or the app token during the upgrade.

I have the feeling that this is related to a zombie session / zombie token, basically something that isn't cleared properly.

**Comment by PVince81** [comment]
And yet another little issue: as I said above, this CSRF check should never kick in for Android users.
We do see a user-agent check in the server code.

However my wife's user-agent says "ownCloudBeta-android/2.5.1-beta.1" but this is the server check:
```
	const USER_AGENT_OWNCLOUD_ANDROID = '/^Mozilla\/5\.0 \(Android\) ownCloud\-android.*$/';
```

I also wonder whether other users in this ticket were using a different or branded version where the user agent did not match ?

So I think we have three issues here:

1) User-Agent check does not correctly match all Android clients: it does not match the beta
2) Some weird in-between authentication/session situation that could cause CSRF for non-Android clients, possibly related to wronglyinvalidated app passwords: https://github.com/owncloud/core/issues/30157
3) Do we really need CSRF for PROPFIND ? I thought it only applies to POST... to be verified

**Comment by DaanTry** [comment]
"I have the feeling that this is related to a zombie session / zombie
token, basically something that isn't cleared properly."

I wonder if that is the case. In my case this is 100% reproducible.

1 Install App
2 Make connection
3 Access files
4 Leave app alone for a few hours to a day (maybe it is the IP switch that
triggers it, going from WLAN to 4G and vice versa)
5 Try to access files
> App reroutes to 'password' screen
6 Login again
> App opens 'files' screen and after a second or so reroutes to  'password'
screen and keeps repeating this behavior after logging in again.

Issue can only be resolved by uninstalling and reinstalling the App. Only
then I can make a connection and access my files again. However the issue
resurfaces after a few hours.

As a workaround I recently discovered that setting 'csrf.disabled' => true
in config.php will make this issue disappear.

**Comment by PVince81** [comment]
@DaanTry are you using the beta client or a branded client ? Are you able to find the user-agent of your client from the access logs ?

From my analysis above, if the Android client is properly detected by the server it should never trigger the code path for CSRF check.

**Comment by davigonz** [comment]
@PVince81 Thank you for the research

@DaanTry is using 2.5.0, as he commented in https://github.com/owncloud/android/issues/1964#issuecomment-353279235.

Other users such as @jensd0e and @pjao reported [here](https://github.com/owncloud/android/issues/2030) the same redirection to login page with 2.5.0 beta 2 and 2.5.1-beta.1 respectively so it's most likely that those redirections appear because as you said above, the user-Agent check doesn't match the beta one and is triggering the CSRF check, which it fails.

**Comment by DaanTry** [comment]
I am using the BETA Android app.

From the server access logs:

xx.xxx.xxx.xx - test [26/Jan/2018:10:15:17 +0100] "PROPFIND
/remote.php/webdav/Data/ HTTP/1.1" 207 23241 "-" "Mozilla/5.0 (Android)
ownCloudBeta-android/2.5.1-beta.1"

**Comment by davigonz** [comment]
> I am using the BETA Android app.

@DaanTry in [this comment](https://github.com/owncloud/android/issues/1964#issuecomment-353279235) you reported the login redirection with 2.5.0, have you noticed the problem with 2.5.0 and after installing 2.5.1-beta.1 it appears again or you just wrote down the wrong app version? Thanks

**Comment by DaanTry** [comment]
I moved to 2.5.1-beta as per your request. Should I retry with the 2.5.0
version?

**Comment by PVince81** [comment]
@DaanTry given the data above, I expect that the problem will disappear for you if you switch back to 2.5.0.

**Comment by DaanTry** [comment]
Update:

- I switched back to Android v2.5.0
- I restored the  'csrf.disabled' in config.php back to false

I have not seen the issue again the past few days.

It seems fixed although I do not understand why and how.

Thanks for all your support!

**Comment by davigonz** [comment]
Hi @johnkj , have you noticed this problem again with the recent versions of the Android app, such as 2.8, released a couple of weeks ago?
